\documentclass[11pt]{article}
\usepackage{enumerate}
\usepackage{fullpage}
\usepackage{fancyhdr}
\usepackage{amsmath, amsfonts, amsthm, amssymb,graphicx,pdfpages}
\setlength{\parindent}{0pt}
\setlength{\parskip}{5pt plus 1pt}
\pagestyle{empty}

\usepackage[utf8]{inputenc}
\usepackage[english]{babel}
\usepackage[english]{isodate}
\usepackage[parfill]{parskip}
\begin{document}
\begin{center}
\textbf{\LARGE{Pipeline for Analysis of Horizontal Gene Transfer in Bacterial Genomes}}\\
~~~\\
\Large{Stuti Agrawal}\\
\Large{Rebecca Elyanow}\\
\Large{Luigi Leung}\\
\Large{Prateek Tandon}\\
\Large{Yiming Xin}
\end{center}
\tableofcontents

\includepdf[pages={1}]{/Users/lleung/Desktop/flowchart.pdf}

\section{Required software}
\begin{verbatim}
mauve Aligner - http://gel.ahabs.wisc.edu/mauve/
jModelTest - https://code.google.com/p/jmodeltest2/downloads/list
Biopython - http://biopython.org/wiki/Main_Page
IPython(Recommended) - http://ipython.org/
Alfy - http://guanine.evolbio.mpg.de/alfy/alfy_1.5.tgz
\end{verbatim}
\section{Running the Pipeline}
The pipeline can be run by calling:\\
\\
aligned\_sequence(data\_dirpath, data\_output, data\_backbone, location\_mauve, location\_jModelTest, output\_model)\\

data\_dirpath = path to directory that holds to sequences you wish to analyze\\
data\_output = the name of the output alignment file from mauve\\
data\_backbone = the name of the output backbone file from mauve\\
location\_mauve = path to mauve program\\
location\_jModelTest = path to jModelTest program\\
output\_model = name of model chosen by jModelTest\\

\section{Alignment}
Use progressive MAUVE \cite{mauve} to align all sequences of the same species in the given directory. The standard defaults of progressive MAUVE are used.\\

\section{Creation of the core genome}
MAUVE’s .backbone file is used to identify regions that are conserved among all genomes and the regions that are not. The regions conserved among all genomes are extracted using this file and concatentated together to create the core genome (including HGTs). \\

\section{Building the phylogenetic tree}
1. Select the best model to fit the phylogenetic tree using JModelTest \cite{jModelTest}\\
2. Generate hundred bootstraps (seqboot) with Phylip \cite{phylip} \\
3. Create a phylogenetic tree using Phylip (default = Maximum Likelihood Tree, DNAML)\\\\
This is done for the core genome (without any HGT regions that are a part
of the core) to identify the evolutionary relationship between the species
and for the pool of the HGT regions that are a part of the core as well as
the distributed genomes so as to identify the relationships between the
organisms with reference to the HGTs (This enables the identification of
the current relationship between the species).\\
The generated phylogenetic trees for the core as well as the HGT regions will be in the phyloxml format.\\

\section{Identifying HGT regions}
Regions of Horizontal Gene Transfer will be identified using the distributed genome of each strain and the program Alfy \cite{Alfy}

\section{Annotate the genes within the horizontally transferred regions}
The sequence regions where HGT has occurred will be queried against the Antibiotic Resistance Gene Database (ARDB),  and for regions that are not in ARDB, they will be queried against the NCBI database.\\

\section{Graphical User Interface}
The GUI is web based. The following are instructions for setting up to locally host the GUI via Apache web server, as well as, instructions for hosting on CMU's AFS server with limited permissions and access.

\subsection{Hosting Locally}
\textbf{Setup}\\
The following instructions apply to OSX and unix-based operating systems. Windows instructions will be in $[$brackets$]$. And instead of \texttt{nano} $[$or notepad$]$, feel free to use any other text editor.\\
On OSX, Apache is already installed. On other unix-based OS, if it is not already installed, install it using the OS's packages utility or via terminal command \texttt{\$ sudo apt-get install apache2}.\\
$[$For Windows, install with the downloaded \texttt{httpd-versionNumber-win32-src.zip}\\
 from \texttt{http://httpd.apache.org/docs/2.2/platform/windows.html}$]$\\
\begin{enumerate}
	\item To get the GUI running on the local computer, please enable php in Apache by going into its httpd.conf by typing into the terminal:
		\begin{verbatim}
		$ sudo nano /etc/apache2/httpd.conf
		
		[ Open C:\\Program Files\Apache Software Foundataion\Apache2.2\
		  and open the httpd.conf in notepad. ]
		\end{verbatim}
	\item Uncomment the line (delete the \# character) , save and exit:
		\begin{verbatim}
		LoadModule php5_module libexec/apache2/libphp5.so
		\end{verbatim}
	\item Enable Apache web server by typing:
		\begin{verbatim}
		$ sudo apachectl start
		(To stop the web server, $ sudo apachectl stop )
			
		[ Click on the httpd.exe in \Apache 2.2\bin\ folder to start the service. ]
		\end{verbatim}
\end{enumerate}
\bigskip

\subsection{Hosting on AFS}
\textbf{Setup}\\
The following instructions apply to hosting on Carnegie Mellon University's server as a student.

To enable CGI and PHP scripts, open in the internet browser:\\
\texttt{https://my.contrib.andrew.cmu.edu/index.cgi}
\smallskip

Enter your AndrewID and password, then under ``\texttt{CGI services}" click on:\\
\texttt{(Re)enable authenticated AFS (CGI AFS-write) support}\\

Upload the .zip contents that you downloaded previously to the \texttt{www} folder in your afs space, or, for your convenience, we have a Github repository for easier uploading and for syncing any possible future updates. To do this, type the following:
\bigskip

Connect to school's clusters (afs):
	\begin{verbatim}
	$ ssh unix.andrew.cmu.edu
	\end{verbatim}
\bigskip

Clone the Github repository to your \texttt{www} directory:
	\begin{verbatim}
	$ git clone git://github.com/713/project.git ~/www/teamB
	\end{verbatim}
\bigskip

\newpage
Give permission to contrib web server to run CGI scripts:
	\begin{verbatim}
		$ cd ~/www/teamB/
		$ fs sa . contrib.[Your AndrewID]@club.cc.cmu.edu rlidwk
	\end{verbatim}
\bigskip

Give write permissions to the CGI, PHP scripts, and output folder:
	\begin{verbatim}
	$ chmod +rw user_email.txt
	$ chmod 755 upload_file.php
	$ chmod 755 user_results
	$ chmod 755 *.cgi
	\end{verbatim}
\bigskip

Edit the location of your server URL in \texttt{config.py}:
	\begin{verbatim}
	$ nano ~/www/teamB/config.py
	\end{verbatim}
\bigskip

Change the ``\texttt{HOST\_713PROJECT=}" variable to:
	\begin{verbatim}
	HOST_713PROJECT="http://www.contrib.cmu.edu/~YourAndrewID/teamB/"
	\end{verbatim}
\subsection{Accessing Web App}

\begin{enumerate}
	\item If hosting locally, open the \texttt{index.html} in browser by dragging it out of its folder and into a browser
	\\or by typing its location in the browser's URL box starting with ``\texttt{file://}"\\
	If hosting on AFS, type into the URL box:
	\begin{verbatim}http://www.contrib.cmu.edu/~YourAndrewID/teamB/
	\end{verbatim}
	\item Select sequences to process.
	\item Provide an email address for receiving an email to view the results when job is finished.\\
		The email is from \texttt{03713.project@gmail.com}\\
		and with the title ``\texttt{Job Completed: A message from 03-713 Team B's web app}"
	\item Click on the blue ``Process" button.
\end{enumerate}
\bigskip

\textbf{Web App Results}\\
After the pipeline is done processing, an email is sent. Within the email, there is a link that redirects the user to the results webpage.

\newpage
\begin{thebibliography}{9}
\bibitem{mauve}
Aaron E. Darling, Bob Mau, and Nicole T. Perna. 2010.  progressiveMauve: Multiple Genome Alignment with Gene Gain, Loss, and Rearrangement.  PLoS One.  5(6):e11147. \\
\bibitem{jModelTest}
Darriba D, Taboada GL, Doallo R, Posada D. 2012. jModelTest 2: more models, new heuristics and parallel computing. Nature Methods 9(8), 772.\\\\
Guindon S and Gascuel O (2003). A simple, fast and accurate method to estimate large phylogenies by maximum-likelihood". Systematic Biology 52: 696-704.\\
\bibitem{phylip}
Felsenstein, J. 1989. PHYLIP - Phylogeny Inference Package (Version 3.2). Cladistics 5: 164-166. \\
\bibitem{Alfy}
Domazet-Lošo M, Haubold B. Alignment-free detection of local similarity among viral and bacterial genomes. Bioinformatics. 2011;27:1466–1472. doi: 10.1093/bioinformatics/btr176.
\end{thebibliography}
\end{document}
